adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,257 @@
+        # * missing: config[:distributed_tracing]
+        # * missing: span.set_error() -- spec level change
+        class TracerMiddleware # rubocop:disable Metrics/ClassLength
+          class << self
+            def allowed_rack_request_headers
+              @allowed_rack_request_headers ||= Array(config[:allowed_request_headers]).each_with_object({}) do |header, memo|
+                memo["HTTP_#{header.to_s.upcase.gsub(/[-\s]/, '_')}"] = build_attribute_name('http.request.headers.', header)
+              end
+            end
+
+            def allowed_response_headers
+              @allowed_response_headers ||= Array(config[:allowed_response_headers]).each_with_object({}) do |header, memo|
+                memo[header] = build_attribute_name('http.response.headers.', header)
+                memo[header.to_s.upcase] = build_attribute_name('http.response.headers.', header)
+              end
+            end
+
+            def build_attribute_name(prefix, suffix)
+              prefix + suffix.to_s.downcase.gsub(/[-\s]/, '_')
+            end
+
+            def config
+              Rack::Adapter.instance.config
+            end
+
+            private
+
+            def clear_cached_config
+              @allowed_rack_request_headers = nil
+              @allowed_response_headers = nil
+            end
+          end
+
+          EMPTY_HASH = {}.freeze
+
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+            original_env = env.dup
+
+            extracted_context = OpenTelemetry.propagation.extract(env)
+
+            # restore extracted context in this process:
+            OpenTelemetry::Context.with_current(extracted_context) do
+              frontend_span = create_frontend_span(env)
+
+              # http://www.rubydoc.info/github/rack/rack/file/SPEC
+              # The source of truth in Rack is the PATH_INFO key that holds the
+              # URL for the current request; but some frameworks may override that
+              # value, especially during exception handling.
+              #
+              # Because of this, we prefer to use REQUEST_URI, if available, which is the
+              # relative path + query string, and doesn't mutate.
+              #
+              # REQUEST_URI is only available depending on what web server is running though.
+              # So when its not available, we want the original, unmutated PATH_INFO, which
+              # is just the relative path without query strings.
+              request_span_name = create_request_span_name(env['REQUEST_URI'] || original_env['PATH_INFO'])
+
+              rack_request = ::Rack::Request.new(env)
+              # Sets as many attributes as are available before control
+              # is handed off to next middleware.
+              request_span = tracer.start_span(request_span_name,
```

Comment:
This makes the frontend_span and request_span siblings, but in dd-trace they are a parent and child. One thing to keep in mind is that `Tracer#trace` and `Tracer#start_span` automatically make the span the active span. For OpenTelemetry `Tracer#start_span` assumes you are explicitly managing context. `Tracer#in_span` and `Tracer#with_span` will activate a span.

Please see my comment again: https://github.com/open-telemetry/opentelemetry-ruby/pull/166#discussion_r376588828

The key points are:
```ruby
#create frontend_span returns a context with the frontend span as the current span
frontend_context = create_frontend_span(env)

#the request span is explicitly parented using the frontend_context 
request_span = tracer.start_span('request span', parent_context: frontend_context || extracted_context)
```

You don't have to use that exact code, but you do need to manage context the same way. Also, it'd be nice to see a test that the frontend span parents the request span in this situation.

Reply:
If this had anything to do with one of the hurdles. let me know and we can look at it.

Reply:
After further thought I suspect this was one of the hurdles. We have a couple of options. The first would be ideal, but would require a change to `in_span` to make it work.

```ruby
def call(env)
  extracted_context = OpenTelemetry.propagation.extract(env)
  
  frontend_context = create_frontend_span(env, extracted_context)
  
  tracer.in_span(..., with_parent_context: frontend_context || extracted_context) do |request_span|
    app.call
  end
  
ensure
  finish_frontend_span(frontend_context)
end

def create_frontend_span(env, extracted_context)
  span = tracer.start_span(..., with_parent_context: extracted_context)
  extracted_context.set_value(current_span_key, span)
end

def finish_frontend_span(context)
  context[current_span_key]&.finish
end
```

For this to work properly, `in_span` would have to have the record and reraise behavior built in. This is relevant to this issue: https://github.com/open-telemetry/opentelemetry-ruby/issues/181.

Without `in_span` having the record and reraise behavior we to drop down to explicitly managed context. Seeing this written down makes me think we might need a few helpers to make explicit management easier:

```ruby
def call(env)
  extracted_context = OpenTelemetry.propagation.extract(env)
  
  frontend_context = create_frontend_span(env, extracted_context)

  request_context = create_request_span(frontend_context || extracted_context)

  OpenTelemetry::Context.with_current(request_context) do
    begin
      @app.call
    rescue => e
      #record and reraise
    ensure
      finish_span(request_context)
      finish_span(frontend_context)
    end
  end
end

def create_frontend_span(env, extracted_context)
  span = tracer.start_span(..., with_parent_context: extracted_context)
  extracted_context.set_value(current_span_key, span)
end

def create_request_span(context)
  span = tracer.start_span(..., with_parent_context: context)
  context.set_value(current_span_key, span)
end

def finish_span(context)
  context[current_span_key]&.finish
end
```

Reply:
@mwear ok, besides the error reporting, I think I got it shape as described now.